Add smith_normal_decomp

isuruf opened the pull request:
This returns `s,t` such that the product `s * m * t` is the
smith_normal_form where `m` is the original matrix
This commit changes the return output type from Matrix to the same
type as of input `m`.

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* matrices
   * Add `smith_normal_decomp` that returns the smith normal form and the matrices
     that decomposes the matrix into the smith normal form.
<!-- END RELEASE NOTES -->

jksuom commented on sympy/matrices/normalforms.py:
```diff
@@ -18,19 +16,34 @@ def smith_normal_form(m, domain = None):
-def invariant_factors(m, domain = None):
+def invariant_factors(m, domain=None, full=False):
     '''
     Return the tuple of abelian invariants for a matrix `m`
```
Perhaps double backticks here as well?

jksuom commented on sympy/matrices/normalforms.py:
```diff
@@ -50,83 +63,111 @@ def invariant_factors(m, domain = None):
         if ind:
             m = m.permute_cols([[0, ind[0]]])
+            if full:
+                t = t.permute_rows([[0, ind[0]]])
```
permute_cols?

jksuom commented on sympy/matrices/tests/test_normalforms.py:
```diff
@@ -1,23 +1,31 @@
 from sympy import Symbol, Poly
-from sympy.polys.solvers import RawMatrix as Matrix
+from sympy.polys.solvers import RawMatrix
```
We could probably ignore this. RawMatrix has been essentially superseded by PolyMatrix that is more flexible.

jksuom commented:
This looks good. It would be possible to make `m.zeros` and `m.eye` use elements of `domain` by setting `m.zero = domain.zero` and `m.one = domain.one`. (Though actually the matrix class, e.g., PolyMatrix, should do that.)

isuruf commented:
> It would be possible to make m.zeros and m.eye use elements of domain by setting m.zero = domain.zero and m.one = domain.one. (Though actually the matrix class, e.g., PolyMatrix, should do that.)

I tried doing that and I ran into the problem I mentioned in gitter about multiplying `PolyElement` and `Poly` types.

jksuom commented:
It should be possible to avoid Poly objects by starting with `R, x  = ring('x', QQ)`. `x` is now an element of `R` (PolyElement), and all expressions like `x**2 - x/2` are also elements of `R`.

jksuom commented:
Then it should be possible to use `R.to_domain()` as `m.ring` (`R` is the low level `PolyRing`). It can also be set when creating a PolyMatrix.

sylee957 commented:
I had some different ideas in terms of design like providing a triplet of API which returns decomposition, normal form, and invariant factors respectively.
Having an option like `full` in `invariant_factors` looks a bit weird.

```diff
$ git diff
diff --git a/sympy/matrices/normalforms.py b/sympy/matrices/normalforms.py
index c41a6e3141..f33be5384b 100644
--- a/sympy/matrices/normalforms.py
+++ b/sympy/matrices/normalforms.py
@@ -1,7 +1,8 @@
 '''Functions returning normal forms of matrices'''
 from __future__ import division, print_function

-def smith_normal_form(m, domain=None, full=False):
+
+def smith_normal_form(m, domain=None):
     '''
     Return the Smith Normal Form of a matrix ``m`` over the ring ``domain``.
     This will only work if the ring is a principal ideal domain.
@@ -29,22 +30,29 @@ def smith_normal_form(m, domain=None, full=False):
     Matrix([[1, 1, 10], [0, -1, -2], [0, 1, 3]])

     '''
-    if full:
-        invs, s, t = invariant_factors(m, domain=domain, full=True)
-    else:
-        invs = invariant_factors(m, domain=domain, full=False)
+    invs = _invariant_factors(m, domain=domain, full=False)
     smf = m.diag(*invs)
     n = len(invs)
     if m.rows > n:
         smf = smf.row_insert(m.rows, m.zeros(m.rows-n, m.cols))
     elif m.cols > n:
         smf = smf.col_insert(m.cols, m.zeros(m.rows, m.cols-n))
-    if full:
-        return smf, s, t
     return smf

-def invariant_factors(m, domain=None, full=False):
-    '''
+
+def smith_decomposition(m, domain=None):
+    invs, s, t = _invariant_factors(m, domain=domain, full=True)
+    smf = m.diag(*invs)
+    n = len(invs)
+    if m.rows > n:
+        smf = smf.row_insert(m.rows, m.zeros(m.rows-n, m.cols))
+    elif m.cols > n:
+        smf = smf.col_insert(m.cols, m.zeros(m.rows, m.cols-n))
+    return smf, s, t
+
+
+def invariant_factors(m, domain=None):
+    """
     Return the tuple of abelian invariants for a matrix ``m``
     (as in the Smith-Normal form)

@@ -53,8 +61,7 @@ def invariant_factors(m, domain=None, full=False):

     [1] https://en.wikipedia.org/wiki/Smith_normal_form#Algorithm
     [2] http://sierra.nmsu.edu/morandi/notes/SmithNormalForm.pdf
-
-    '''
+    """
     if not domain:
         if not (hasattr(m, "ring") and m.ring.is_PID):
             raise ValueError(
@@ -62,6 +69,10 @@ def invariant_factors(m, domain=None, full=False):
         else:
             domain = m.ring

+    return _invariant_factors(m, domain=domain, full=False)
+
+
+def _invariant_factors(m, domain=None, full=False):
     if len(m) == 0:
         if full:
             return (), m, m
@@ -158,8 +169,8 @@ def clear_row(m, s, t):
         invs = ()
     else:
         if full:
-            invs, s_small, t_small = invariant_factors(m[1:,1:], domain=domain,
-                                        full=True)
+            invs, s_small, t_small = \
+                _invariant_factors(m[1:,1:], domain=domain, full=True)
             s2 = m.eye(m.rows)
             t2 = m.eye(m.cols)
             s2[1:,1:] = s_small
@@ -167,7 +178,7 @@ def clear_row(m, s, t):
             s = s2 * s
             t = t * t2
         else:
-            invs = invariant_factors(m[1:,1:], domain=domain, full=False)
+            invs = _invariant_factors(m[1:,1:], domain=domain, full=False)

     if m[0,0]:
         result = [m[0,0]]
(END)
```

czgdp1807 commented:
@isuruf Would you like to complete this PR?

isuruf commented:
Yes, when I get some time.

sylee957 commented:
Now, since DomainMatrix is available, the changes should better be exported to DomainMatrix and Matrix can be converted to DomainMatrix to compute it.

oscarbenjamin commented:
I've opened a PR #21402 that moves the `smith_normal_form` function to `polys` and converts it to use `DomainMatrix`.

smichr commented:
This is a modification of `invariant_factors` that produces the decomposition:
```python
def smith_decomp(m):
    """return A, S, T for domain matrix m such that
    S*m*T is a diagonal matrix with diagonal A"""
    domain = m.domain
    if not domain.is_PID:
        raise ValueError('all entries must be over PID')
    if 0 in m.shape:
        return (), Matrix(), Matrix()
    rows, cols = shape = m.shape
    s = list(m.eye(rows,m.domain).to_dense().rep.to_ddm())
    t = list(m.eye(cols,m.domain).to_dense().rep.to_ddm())
    # make copy of m
    m = [[c for c in r] for r in list(m.to_dense().rep.to_ddm())]
    def add_columns(m, i, j, a, b, c, d):
        for _ in (m,t):
            for k in range(len(_)):
                e = _[k][i]
                _[k][i] = a*e + b*_[k][j]
                _[k][j] = c*e + d*_[k][j]
    def add_rows(m, i, j, a, b, c, d):
        for _ in (m,s):
            for k in range(len(_[0])):
                e = _[i][k]
                _[i][k] = a*e + b*_[j][k]
                _[j][k] = c*e + d*_[j][k]
    def clear_column(m):
        if m[0][0] == 0:
            return m  # pragma: nocover
        pivot = m[0][0]
        for j in range(1, rows):
            if m[j][0] == 0:
                continue
            d, r = domain.div(m[j][0], pivot)
            if r == 0:
                add_rows(m, 0, j, 1, 0, -d, 1)
            else:
                a, b, g = domain.gcdex(pivot, m[j][0])
                d_0 = domain.div(m[j][0], g)[0]
                d_j = domain.div(pivot, g)[0]
                add_rows(m, 0, j, a, b, d_0, -d_j)
                pivot = g
        return m
    def clear_row(m):
        if m[0][0] == 0:
            return m  # pragma: nocover
        pivot = m[0][0]
        for j in range(1, cols):
            if m[0][j] == 0:
                continue
            d, r = domain.div(m[0][j], pivot)
            if r == 0:
                add_columns(m, 0, j, 1, 0, -d, 1)
            else:
                a, b, g = domain.gcdex(pivot, m[0][j])
                d_0 = domain.div(m[0][j], g)[0]
                d_j = domain.div(pivot, g)[0]
                add_columns(m, 0, j, a, b, d_0, -d_j)
                pivot = g
        return m
    # swap rows or columns so m[0,0] is non-zero if possible
    ind = [i for i in range(rows) if m[i][0] != 0]
    if ind and (i:=ind[0]) != 0:
        for _ in (m,s):
            _[0], _[i] = _[i], _[0]
    else:
        ind = [j for j in range(cols) if m[0][j] != 0]
        if ind and (i:=ind[0]) != 0:
            for _ in (m,t):
                for row in _:
                    row[0], row[i] = row[i], row[0]
    # make the first row and column except m[0,0] zero
    while (any(m[0][i] != 0 for i in range(1,cols)) or
           any(m[i][0] != 0 for i in range(1,rows))):
        m = clear_column(m)
        m = clear_row(m)
    if 1 in shape:
        invs = ()
    else:
        assert all(type(i) is list for i in (m,s,t))
        view = DomainMatrix([r[1:] for r in m[1:]],
        (rows-1,cols-1), domain)
        invs,si,ti = smith_decomp(view)
        s,t,si,ti = [Matrix(i) for i in (s,t,si,ti)]
        s2 = Matrix.eye(rows)
        t2 = Matrix.eye(cols)
        s2[1:,1:] = si
        t2[1:,1:] = ti
        s = s2 * s
        t = t * t2
    if m[0][0]:
        result = [m[0][0]]
        result.extend(invs)
        for i in range(len(result)-1):
            if result[i] and domain.div(result[i+1], result[i])[1] != 0:
                g = domain.gcd(result[i+1], result[i])
                p = domain.div(result[i], g)[0]
                s[i+1,:]*=p
                s[i,:]*=g/result[i]
                result[i+1] = p*result[i+1]
                result[i] = g
            else:
                break
    else:
        s, t = [Matrix(i) for i in (s,t)]
        s = s.permute_rows([[i,i+1] for i in range(s.rows-1)])
        t = t.permute_cols([[i,i+1] for i in range(t.cols-1)])
        result = invs + (m[0][0],)
    return tuple(result),s,t
```
Example:
```python
>>> m=randMatrix(3,5)
>>>A,S,T=smith_decomp(DMZZ(m))
>>> m
Matrix([
[42,  0, 10, 33, 13],
[76,  8, 32, 77, 93],
[19, 56, 67, 82, 68]])
>>> S*m*T
Matrix([
[1, 0, 0, 0, 0],
[0, 1, 0, 0, 0],
[0, 0, 4, 0, 0]])
>>> A
(1, 1, 4)

from sympy.polys.matrices import DomainMatrix
from sympy import ZZ,Matrix,diag,randMatrix
def check(m):
    A, S, T = smith_decomp(m)
    assert tuple((S*m*T).diagonal().tolist()[0])==A

def DMZZ(m):
    if isinstance(m, Matrix):
        m = m.tolist()
    if type(m) == list:
        m = [[ZZ(i) for i in r] for r in m]
        shape = len(m),len(m[0])
        return DomainMatrix(m, shape, ZZ)
    assert None, 'unrecognized'

check(DMZZ([[12,6,4],[3,9,6],[2,16,14]]))
m= Matrix([[1,-1,1],[2,-2,2]])
check(DMZZ(m))
check(DMZZ(m.T))
```

oscarbenjamin commented:
It is a lot easier to convert to DomainMatrix now when wanted:
```python
In [4]: M = Matrix([[1, 2], [3, 4]])

In [5]: M.to_DM()
Out[5]: DomainMatrix({0: {0: 1, 1: 2}, 1: {0: 3, 1: 4}}, (2, 2), ZZ)

In [6]: M.to_DM(QQ)
Out[6]: DomainMatrix({0: {0: 1, 1: 2}, 1: {0: 3, 1: 4}}, (2, 2), QQ)

In [7]: M.to_DM(QQ).to_Matrix()
Out[7]: 
⎡1  2⎤
⎢    ⎥
⎣3  4⎦
```

isuruf committed:
Add smith_normal_decomp

This returns the smith normal form and the matrices that decomposes
the matrix into the smith normal form

isuruf committed:
fix doctest

isuruf committed:
Keep using lists

isuruf committed:
simplify code and fix test

oscarbenjamin commented:
I tested with this:
```python
from sympy import *

from sympy.matrices.normalforms import smith_normal_decomp

for i in range(6):
    for j in range(6):
        for _ in range(10 if i*j else 1):
            m = randMatrix(i, j, max=5, percent=50)
            try:
                a, s, t = smith_normal_decomp(m)
                assert a == s * m * t
            except Exception as e:
                print(m)
                print(e)
```
There were a bunch of failures:
```python
Matrix(0, 1, [])
Inconsistent row-list/shape
Matrix(0, 2, [])
Inconsistent row-list/shape
Matrix(0, 3, [])
Inconsistent row-list/shape
Matrix(0, 4, [])
Inconsistent row-list/shape
Matrix(0, 5, [])
Inconsistent row-list/shape
Matrix(1, 0, [])
Inconsistent row-list/shape
Matrix(2, 0, [])
Inconsistent row-list/shape
Matrix([[0, 3], [4, 0]])
'int' object is not iterable
Matrix([[4, 0], [0, 1]])
'int' object is not iterable
Matrix([[4, 0, 0], [0, 4, 5]])
'int' object is not iterable
Matrix([[0, 0, 1], [5, 0, 0]])
'int' object is not iterable
Matrix([[5, 0, 0, 5], [5, 0, 1, 0]])
'int' object is not iterable
Matrix([[0, 0, 0, 5], [3, 3, 0, 0]])
'int' object is not iterable
Matrix(3, 0, [])
Inconsistent row-list/shape
Matrix([[5, 0], [0, 0], [0, 2]])
'int' object is not iterable
Matrix([[0, 3], [3, 0], [0, 2]])
'int' object is not iterable
Matrix([[2, 0, 0], [0, 2, 0], [4, 0, 3]])
'int' object is not iterable
Matrix([[0, 0, 0], [0, 0, 2], [4, 0, 0]])
'int' object is not iterable
Matrix([[5, 0, 0], [0, 0, 2], [3, 4, 0]])
'int' object is not iterable
Matrix([[1, 2, 0], [5, 0, 0], [0, 0, 3]])
'int' object is not iterable
Matrix([[0, 3, 0], [5, 0, 0], [0, 4, 0]])
'int' object is not iterable
Matrix([[4, 0, 0], [0, 3, 1], [0, 1, 0]])
'int' object is not iterable
Matrix([[4, 0, 0, 0], [0, 0, 4, 5], [0, 2, 2, 4]])
'int' object is not iterable
Matrix([[2, 2, 0, 0], [0, 3, 0, 1], [0, 3, 3, 0]])
'int' object is not iterable
Matrix([[0, 2, 0, 1], [5, 0, 0, 5], [5, 0, 0, 3]])
'int' object is not iterable
Matrix([[3, 3, 0, 0], [1, 0, 0, 0], [2, 0, 0, 1]])
'int' object is not iterable
Matrix([[4, 0, 0, 0], [0, 0, 5, 0], [4, 1, 0, 3]])
'int' object is not iterable
Matrix([[4, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 4]])
'int' object is not iterable
Matrix([[0, 3, 3, 0], [0, 0, 0, 1], [0, 3, 0, 5]])
'int' object is not iterable
Matrix([[0, 4, 5, 1, 3], [3, 3, 0, 0, 0], [0, 0, 1, 0, 0]])
'int' object is not iterable
Matrix([[0, 0, 1, 0, 4], [0, 0, 0, 0, 0], [2, 2, 0, 4, 2]])
'int' object is not iterable
Matrix([[2, 1, 3, 0, 0], [0, 0, 0, 2, 4], [0, 0, 3, 0, 0]])
'int' object is not iterable
Matrix([[5, 3, 3, 0, 0], [2, 0, 0, 0, 0], [0, 0, 0, 4, 4]])
'int' object is not iterable
Matrix(4, 0, [])
Inconsistent row-list/shape
Matrix([[3, 0], [3, 5], [0, 4], [0, 0]])
'int' object is not iterable
Matrix([[0, 0], [3, 0], [0, 1], [0, 1]])
'int' object is not iterable
Matrix([[0, 2, 0], [0, 4, 0], [2, 0, 4], [0, 1, 0]])
'int' object is not iterable
Matrix([[4, 0, 4], [0, 4, 0], [0, 5, 4], [0, 4, 0]])
'int' object is not iterable
Matrix([[5, 0, 0], [0, 0, 4], [0, 0, 0], [0, 0, 0]])
'int' object is not iterable
Matrix([[4, 0, 0], [2, 0, 0], [0, 1, 2], [0, 0, 3]])
'int' object is not iterable
Matrix([[1, 3, 4, 0], [0, 0, 2, 0], [0, 0, 0, 0], [0, 3, 0, 3]])
'int' object is not iterable
Matrix([[5, 0, 0, 0], [5, 1, 0, 0], [0, 1, 2, 2], [0, 0, 1, 5]])
'int' object is not iterable
Matrix([[2, 4, 0, 0, 0], [2, 0, 2, 0, 0], [0, 0, 0, 1, 2], [4, 3, 0, 3, 0]])
'int' object is not iterable
Matrix([[0, 0, 1, 0, 0], [0, 0, 1, 0, 4], [0, 0, 5, 0, 0], [2, 0, 2, 0, 0]])
'int' object is not iterable
Matrix([[1, 0, 5, 0, 0], [0, 0, 0, 5, 0], [2, 4, 4, 0, 4], [5, 0, 5, 0, 0]])
'int' object is not iterable
Matrix([[0, 0, 2, 0, 0], [4, 5, 0, 3, 0], [3, 3, 0, 5, 0], [0, 0, 0, 2, 5]])
'int' object is not iterable
Matrix(5, 0, [])
Inconsistent row-list/shape
Matrix([[0, 0], [4, 0], [0, 0], [0, 3], [2, 2]])
'int' object is not iterable
Matrix([[0, 5, 1], [0, 5, 0], [4, 5, 0], [0, 4, 0], [0, 1, 0]])
'int' object is not iterable
Matrix([[0, 4, 0], [4, 0, 0], [0, 0, 0], [0, 0, 0], [0, 3, 4]])
'int' object is not iterable
Matrix([[4, 0, 2], [2, 0, 0], [0, 0, 0], [2, 5, 0], [0, 0, 2]])
'int' object is not iterable
Matrix([[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 2, 0, 5], [2, 0, 4, 0]])
'int' object is not iterable
Matrix([[1, 2, 2, 0], [3, 2, 0, 0], [0, 4, 1, 0], [0, 0, 3, 0], [0, 0, 0, 5]])
'int' object is not iterable
Matrix([[1, 4, 2, 0], [5, 4, 4, 0], [0, 0, 0, 4], [0, 0, 4, 0], [0, 0, 4, 0]])
'int' object is not iterable
Matrix([[0, 2, 0, 0, 4], [0, 1, 0, 0, 0], [0, 0, 0, 2, 2], [4, 5, 5, 0, 3], [0, 0, 0, 0, 3]])
'int' object is not iterable
Matrix([[5, 0, 0, 5, 1], [2, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 4, 4, 0, 0], [0, 4, 0, 1, 3]])
'int' object is not iterable
Matrix([[0, 0, 3, 0, 4], [0, 0, 1, 0, 0], [0, 2, 0, 5, 0], [0, 1, 0, 0, 2], [0, 0, 5, 5, 0]])
'int' object is not iterable
```

oscarbenjamin commented:
Also:
```python
In [1]: from sympy.matrices.normalforms import smith_normal_decomp

In [2]: M = randMatrix(3) + randMatrix(3)*x

In [3]: M
Out[3]: 
⎡49⋅x + 8   93⋅x + 28  83⋅x + 84⎤
⎢                               ⎥
⎢3⋅x + 18   68⋅x + 42  8⋅x + 18 ⎥
⎢                               ⎥
⎣92⋅x + 83  14⋅x + 51  79⋅x + 89⎦

In [4]: smith_normal_decomp(M)
---------------------------------------------------------------------------
ValueError 
...
ValueError: The matrix entries must be over a principal ideal domain
> /home/oscar/current/active/sympy/sympy/polys/matrices/normalforms.py(103)_smith_normal_decomp()
    101     if not domain.is_PID:
    102         msg = "The matrix entries must be over a principal ideal domain"
--> 103         raise ValueError(msg)
    104 
    105     if len(m) == 0 or len(m[0]) == 0:

ipdb> p domain
ZZ[x]
...
```
So `ZZ[x]` is not a PID but `QQ[x]` is:
```python
In [26]: m = randMatrix(3) + randMatrix(3)*x

In [27]: print(m)
Matrix([[38*x + 23, 35*x + 75, 3*x + 47], [69*x + 87, 30*x + 50, 19*x + 80], [45*x + 26, 91*x + 87, 26*x + 93]])

In [28]: a, s, t = smith_normal_decomp(m, QQ[x])

In [29]: simplify(s * m * t) == a
Out[29]: True

In [30]: a
Out[30]: 
⎡1  0                      0                     ⎤
⎢                                                ⎥
⎢0  1                      0                     ⎥
⎢                                                ⎥
⎢               3           2                    ⎥
⎣0  0  - 54140⋅x  - 172428⋅x  - 222792⋅x - 209312⎦

In [31]: s
Out[31]: 
⎡          -23                        38                                       ⎤
⎢          ────                      ────                         0            ⎥
⎢          573                       1719                                      ⎥
⎢                                                                              ⎥
⎢103021029⋅x   236449587     39356583⋅x    26659933     5329755⋅x     23991741 ⎥
⎢─────────── + ──────────  - ────────── - ──────────  - ────────── - ──────────⎥
⎢5964315040    5964315040    5964315040   5964315040    1192863008   1192863008⎥
⎢                                                                              ⎥
⎢      2                            2                         2                ⎥
⎣4929⋅x  + 10890⋅x + 6269   - 1883⋅x  - 1114⋅x - 51   - 1275⋅x  - 5630⋅x - 5375⎦

In [32]: t
Out[32]: 
⎡                             3                2                             ⎤
⎢   425⋅x   3275  8015359325⋅x    46803822415⋅x    32842492325⋅x   5789290483⎥
⎢1  ───── + ────  ───────────── + ────────────── + ───────────── + ──────────⎥
⎢    573    1719   56959208632     56959208632      28479604316    7119901079⎥
⎢                                                                            ⎥
⎢                                      2                                     ⎥
⎢                            56579007⋅x    462623831⋅x   67994488            ⎥
⎢0       1                   ─────────── + ─────────── + ─────────           ⎥
⎢                             298215752     745539380    186384845           ⎥
⎢                                                                            ⎥
⎣0       0                                     1                             ⎦
```
It would probably be good to have a `to_PID()` method for domains. Alternatively here we could just check for a univariate polynomial ring and try to convert the domain to a field.

This is not right:
```python
In [33]: M = randMatrix(3) + randMatrix(3)*I

In [34]: smith_normal_decomp(M)
---------------------------------------------------------------------------
..
ValueError: The matrix entries must be over a principal ideal domain
```
Here `ZZ_I` is a PID:
```python
In [35]: ZZ_I.is_PID # wrong
Out[35]: False
```

isuruf committed:
Fix the doctest and some tests

isuruf committed:
Add more tests and fix the algorithm

isuruf committed:
remove impossible if condition

isuruf commented on sympy/polys/matrices/normalforms.py:
```diff
@@ -59,83 +59,157 @@ def invariant_factors(m):
         # make m[1:, 0] zero by row and column operations
         if m[0][0] == 0:
-            return m  # pragma: nocover
+            return  # pragma: nocover
```
It wouldn't. Removed this if condition.

oscarbenjamin replied:
It looks like you actually removed the other place where this happens. I guess this one can go as well.

isuruf commented on sympy/matrices/normalforms.py:
```diff
@@ -43,6 +44,25 @@ def smith_normal_form(m, domain=None):
+    >>> from sympy.matrices.normalforms import smith_normal_decomp
+    >>> m = Matrix([[12, 6, 4], [3, 9, 6], [2, 16, 14]])
+    >>> a, s, t = smith_normal_decomp(m, domain=ZZ)
+    >>> assert a == m * s * t
```
Fixed

oscarbenjamin commented on sympy/polys/matrices/normalforms.py:
```diff
@@ -59,98 +59,194 @@ def invariant_factors(m):
+                    add_columns(t, 0, j, 1, 0, -d, 1)
             else:
                 a, b, g = domain.gcdex(pivot, m[0][j])
                 d_0 = domain.div(m[0][j], g)[0]
                 d_j = domain.div(pivot, g)[0]
```
Is this intended to be exact division or is there a nonzero remainder that is being discarded?

If so it is typically better to use `domain.exquo`.

isuruf replied:
Using `exquo` now

isuruf committed:
Use exquo and remove another redundant condition

oscarbenjamin commented:
If you have python-flint installed then you can compare with Flint e.g.:
```python
In [14]: M = randMatrix(3, 4)

In [15]: M
Out[15]: 
⎡59  16  56  95⎤
⎢              ⎥
⎢24  93  21  79⎥
⎢              ⎥
⎣61  76  47  76⎦

In [16]: smith_normal_decomp(M)
Out[16]: 
⎛                                     ⎡1  2335  28129   170169 ⎤⎞
⎜⎡1  0  0  0⎤  ⎡ 11     -27     0  ⎤  ⎢                        ⎥⎟
⎜⎢          ⎥  ⎢                   ⎥  ⎢0   1    -4505   -28282 ⎥⎟
⎜⎢0  1  0  0⎥, ⎢1682   1533   -2230⎥, ⎢                        ⎥⎟
⎜⎢          ⎥  ⎢                   ⎥  ⎢0   0    -40616  -254959⎥⎟
⎜⎣0  0  1  0⎦  ⎣-3849  -3508  5103 ⎦  ⎢                        ⎥⎟
⎝                                     ⎣0   0     7865    49371 ⎦⎠

In [17]: M.to_DM().to_dense().rep.rep.snf()
Out[17]: 
[1, 0, 0, 0]
[0, 1, 0, 0]
[0, 0, 1, 0]
```
I adjusted the test script to compare with python-flint:
```python
from sympy import *

from sympy.matrices.normalforms import smith_normal_decomp

for i in range(1, 6):
    for j in range(1, 6):
        print(i, j)
        for _ in range(10 if i*j else 1):
            m = randMatrix(i, j, max=5, percent=50)
            a, s, t = smith_normal_decomp(m)
            assert a == s * m * t
            a2 = m.to_DM().to_dense().rep.rep.snf()
            a2 = Matrix(a2.tolist())
            if a != a2:
                print(m)
                print(a)
                print(a2)
            else:
                print('equal!')
```
There seem to be lots of cases where they are different:
```python
In [33]: M = Matrix([[0, 3, 4], [0, 0, 4]])

In [34]: M
Out[34]: 
⎡0  3  4⎤
⎢       ⎥
⎣0  0  4⎦

In [35]: smith_normal_decomp(M)
Out[35]: 
⎛                      ⎡0   0   1⎤⎞
⎜⎡1   0   0⎤  ⎡-3  1⎤  ⎢         ⎥⎟
⎜⎢         ⎥, ⎢     ⎥, ⎢-1  -8  0⎥⎟
⎜⎣0  -12  0⎦  ⎣-4  1⎦  ⎢         ⎥⎟
⎝                      ⎣1   9   0⎦⎠

In [36]: M.to_DM().to_dense().rep.rep.snf()
Out[36]: 
[1,  0, 0]
[0, 12, 0]
```
I think the differences are all just minus signs. It looks like Flint's implementation of SNF only ever has positive numbers.

According to wikipedia the invariant factors are unique up to multiplication by a unit so for integers that means `+-1`. The domain system has a special method for this called `canonical_unit`:
```python
In [38]: ZZ.canonical_unit(ZZ(-6))
Out[38]: -1

In [39]: ZZ_I.canonical_unit(ZZ_I(1,-2))
Out[39]: ZZ_I(0, 1)

In [42]: R = QQ[x]

In [43]: R.canonical_unit(R(-x))
Out[43]: -1

In [44]: R = QQ_I[x]

In [45]: R.canonical_unit(R(I*x + 1))
Out[45]: (0 + -1*I)
```
This is used for canonicalising the signs of things like polynomial gcd or the denominator of a rational function. It isn't really documented anywhere but the idea is that if `u = K.canonical_unit(e)` then `u*e` is canonical with respect to multiplication by a basic group of units:
```python
In [51]: canonicalise = lambda K, x: K.canonical_unit(x)*x

In [52]: [canonicalise(ZZ_I, ZZ_I.from_sympy(e)) for e in [S(1), -S(1), I, -I]]
Out[52]: [ZZ_I(1, 0), ZZ_I(1, 0), ZZ_I(1, 0), ZZ_I(1, 0)]
```
Is it possible to incorporate that into the choice of invariant factors so that the output matches Flint?

It is particularly useful to match Flint because ideally this would be a `DomainMatrix.snf()` method and then the implementation here would be for `DDM` but we could use Flint's implementation for at least the SNF when Flint is installed. (Flint doesn't have the full decomposition.)

isuruf commented:
@oscarbenjamin I think all your suggestions are valid, but they are all defects in the existing `smith_normal_form` and should probably be separate issues.

oscarbenjamin commented:
Some of my points about using non-integer domains are more general comments about the domain system rather than things to be fixed here.

At some point I had to translate the SNF code from using RawMatrix to using domain elements and DomainMatrix because we needed to get rid of RawMatrix. It was clear to me at the time that there were problems with the code but I didn't understand the algorithm well enough to fix it. It would have been great if I could have talked to someone who actually understood the algorithm to fix these things. That is why I am taking the opportunity now to ask you about them since you seem to know this algorithm.

If we can get to the point that correctness if verifiable for a large number of random inputs and the implementation is well tested then it will be possible for others to improve this code in future. Ideally this whole implementation would be rewritten as a DDM method but I don't want to touch it while it is broken.

It is very unclear to me where in the algorithm the `K.canonical_unit` method would be used to make the signs of the SNF canonical. A simple fix is just to do it at the end which I guess just means flipping the signs in `A` and then flipping the corresponding signs in rows or columns of `S` and `T`.

I'm assuming though that there is a natural place in the algorithm where we can just flip the sign of one or two things like:
```python
if a < 0:
   a, b = -a, -b
   if full:
       x, y = -x, -y
```
I can't see it though because I don't know the algorithm well enough to follow the code. The changes here now make it more complicated to fix it because it presumably also requires adjusting the code in the `full=True` code paths as well.

If we have
```
a == s * m * t
```
Then if `u` is a diagonal matrix of units that fixes the signs in `a` we have:
```
(u * a) == (u * s) * m * t
a' == s' * m * t
```
So I guess somewhere we can fix the sign of `a[i,i]` but we need to multiply row `s[i]` by the same factor or something like that. Maybe it is just easiest to do that at the end:
```diff
diff --git a/sympy/polys/matrices/normalforms.py b/sympy/polys/matrices/normalforms.py
index 95695309e7..804bf7e8ca 100644
--- a/sympy/polys/matrices/normalforms.py
+++ b/sympy/polys/matrices/normalforms.py
@@ -89,6 +89,12 @@ def smith_normal_decomp(m):
 
     s = DomainMatrix(s, domain=domain, shape=(rows, rows))
     t = DomainMatrix(t, domain=domain, shape=(cols, cols))
+
+    units = [domain.canonical_unit(i) for i in invs]
+    u = DomainMatrix.diag(units, domain, (rows, rows))
+    smf = u*smf
+    s = u*s
+
     return smf, s, t


```
I still see differences though. It is not just about the minus signs...

Here is the simplest example with this PR:
```python
In [4]: M = Matrix([[4, 0], [0, 2]])

In [6]: M.to_DM().to_dense().rep.rep.snf()
Out[6]: 
[2, 0]
[0, 4]

In [7]: smith_normal_decomp(M)[0]
Out[7]: 
⎡4  0⎤
⎢    ⎥
⎣0  2⎦
```
That is definitely incorrect because 4 does not divide 2. It seems to be that the fact that 2 divides 4 is significant here in causing the wrong result.

Actually on this is a change brought by the PR. On master we have this which is correct:
```python
In [3]: smith_normal_form(Matrix([[4, 0], [0, 2]]))
Out[3]: 
⎡2  0⎤
⎢    ⎥
⎣0  4⎦
```
The PR gives:
```python
In [2]: smith_normal_form(Matrix([[4, 0], [0, 2]]))
Out[2]: 
⎡4  0⎤
⎢    ⎥
⎣0  2⎦
```
We also need the tests to assert that the output actually is in Smith normal form like `assert is_snf(M)`.

isuruf committed:
Fix normal form and add checks

oscarbenjamin commented:
Okay my testing suggests that this is correct up to signs now:
```python
from sympy import *

from sympy.matrices.normalforms import smith_normal_decomp

for i in range(6):
    for j in range(6):
        print(i, j)
        for _ in range(1000 if i*j else 1):
            m = randMatrix(i, j, max=5, percent=50)
            a, s, t = smith_normal_decomp(m)
            assert a == s * m * t
            a2 = a3 = m.to_DM().to_dense().rep.rep.snf()
            #print(m)
            #print(a)
            #print(a2)
            assert a.applyfunc(abs).to_DM().to_dense().rep.rep == a2
```

isuruf commented:
> So I guess somewhere we can fix the sign of a[i,i] but we need to multiply row s[i] by the same factor or something like that. Maybe it is just easiest to do that at the end:

Yep, that's right, but not at the end. We want to do this in `_smith_normal_decomp` so that everything is consistent.

Following would make it work.
```diff
diff --git a/sympy/polys/matrices/normalforms.py b/sympy/polys/matrices/normalforms.py
index 25ae2a17c5..6b3a80cf71 100644
--- a/sympy/polys/matrices/normalforms.py
+++ b/sympy/polys/matrices/normalforms.py
@@ -216,6 +216,13 @@ def clear_row():
         clear_column()
         clear_row()
 
+    if m[0][0] != 0:
+        c = domain.canonical_unit(m[0][0])
+        if c != domain.one:
+            m[0][0] *= c
+            if full:
+                s[0] = [elem * c for elem in s[0]]
+
     def to_domain_matrix(m):
         return DomainMatrix(m, shape=(len(m), len(m[0])), domain=domain)
```

isuruf commented:
> So ZZ[x] is not a PID but QQ[x]
> It would probably be good to have a to_PID() method for domains. Alternatively here we could just check for a univariate polynomial ring and try to convert the domain to a field.

True. The automatic domain inference needs to be better.

> Also QQ_I is a PID but doesn't have gcdex:

implementing `gcdex` for `QQ_I` should be easy enough.

Bezout's Identity (https://en.wikipedia.org/wiki/B%C3%A9zout%27s_identity) holds in every PID, so a `gcdex` should exist although it might not be possible to find `gcdex` simply by using Euclidean algorithm. (I think  all the domains we have domains in SymPy that is a PID, it is possible to find `gcdex` by using Euclidean algorithm)

oscarbenjamin commented:
> > Also QQ_I is a PID but doesn't have gcdex:
> 
> implementing `gcdex` for `QQ_I` should be easy enough.

Actually I'm not sure this makes sense. `QQ_I` is a field like `QQ` which also does not have gcdex:
```python
In [7]: QQ.gcdex(QQ(3), QQ(5))
---------------------------------------------------------------------------
NotImplementedError
```
But I can trick this:
```python
In [8]: R = QQ[x]

In [9]: R.gcdex(R(3), R(5))
Out[9]: (0, 1/5, 1)
```
I suppose implementing `gcdex` for fields is kind of trivial if we just want to denote `1` as the gcd of everything.

Is the SNF over a field basically equivalent to something else?

oscarbenjamin committed:
fix(matrices): make DDM.to_list() copy the lists

oscarbenjamin committed:
fix(matrices): fix signs in smith_normal_form

isuruf commented:
> Is the SNF over a field basically equivalent to something else?

Yes. If the matrix is square and has full rank then, it's just `I = M^-1 M I`. A simple LU based algorithm can give the SNF.

oscarbenjamin committed:
test(matrices): add tests for smith_normal_form

oscarbenjamin commented:
I've just pushed a bunch of changes:

1. Fix the list copying bug.
2. Fix the signs of SNF so it matches Flint.
3. Added tests.

@isuruf do you want to check it over?

isuruf commented:
All your commits look good to me.

oscarbenjamin commented:
> The automatic domain inference needs to be better.

This particular case is a bit special because we want to try to interpret the expressions as elements of a PID but *not* a field given that SNF is not particularly meaningful over a field.

I think what that means then is that if we have a univariate polynomial ring then we want to set the ground domain to a field. Otherwise the only casess we could handle reasonably would be ZZ or ZZ_I and those would be detected automatically (the `ZZ_I.is_pid` bug should be fixed though). Anything else would just have to become a field and produce a somewhat less meaningful result.

Perhaps it doesn't make much sense to even talk about SNF unless the user specifies a domain although we should pick the obvious common cases: `ZZ`, `ZZ_I`, and `K[x]` for a field `K`.

> I'm not sure when we can't find `gcdex` using Euclidean algorithm

It is when we don't have Euclidean division (or cannot reasonably compute it). SymPy every PID is either a field or has Euclidean division so there aren't any obvious counter examples.

isuruf committed:
Add an easy out for fields

oscarbenjamin commented:
Thanks for this @isuruf I will merge this when the tests pass.

The `smith_normal_form` routine is looking in good shape. It is properly tested with 100% coverage and corner cases handled etc. I'm much happier that this is correct now than I was when I last looked at the code. Computing the full decomposition is a nice feature that does not seem to be otherwise available in Python outside of Sage.

isuruf committed:
test that s and t are invertible and in the same domain

oscarbenjamin commented:
We should implement `gcdex` for all fields just for completeness.

I'm trying to think how the LU approach works. Generally the idea with LU is that we start with M and then we compute
```
P*M = L*U
```
where `P` is an invertible permutation matrix. Here with the SNF we start with M and compute
```
S*M*T = A
```
where `S` and `T` are invertible and `A` is diagonal. I guess if L and U are invertible then it is just:
```
L^-1*P*M*U^-1 = I
```
Then we have `S = L^-1*P`, `T = U^-1` and `A` = `I`.

Generally the LU is defined so that L is invertible but U is not necessarily e.g.:
```
In [3]: M = Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]])

In [4]: M.LUdecomposition()
Out[4]: 
⎛⎡1  0  0⎤  ⎡1  2   3 ⎤    ⎞
⎜⎢       ⎥  ⎢         ⎥    ⎟
⎜⎢4  1  0⎥, ⎢0  -3  -6⎥, []⎟
⎜⎢       ⎥  ⎢         ⎥    ⎟
⎝⎣7  2  1⎦  ⎣0  0   0 ⎦    ⎠
```
So we can always comute `L^-1` but not necessarily `U^-1`. However I guess we can factorise `U` into a product of an invertible matrix and a SNF like diagonal matrix e.g.:
```python
In [8]: A = Matrix.diag(1, 1, 0)

In [9]: Up = Matrix([[1, 2, 3], [0, -3, -6], [0, 0, 1]])

In [12]: Eq(MatMul(A, Up), A*Up)
Out[12]: 
⎡1  0  0⎤ ⎡1  2   3 ⎤   ⎡1  2   3 ⎤
⎢       ⎥ ⎢         ⎥   ⎢         ⎥
⎢0  1  0⎥⋅⎢0  -3  -6⎥ = ⎢0  -3  -6⎥
⎢       ⎥ ⎢         ⎥   ⎢         ⎥
⎣0  0  0⎦ ⎣0  0   1 ⎦   ⎣0  0   0 ⎦
```
Now `U = A*Up` where `Up` is invertible so we have:
```python
P*M = L*U
P*M = L*A*Up
L^-1*P*M*Up^-1 = A
```
Then `S = L^-1*P`, `T = Up^-1` and `A = A`.

So that's the square case...

Still in the rectangular case we have `L` as square and invertible. But `U` is rectangular and not invertible. I assume we can still write it as `A*Up` for diagonal `A` and square invertible `Up` and then it is the same as for the square case...

There is a naive implementation of `LU` decomposition for DomainMatrix over a field domain:
```python
In [26]: M = randMatrix(3, 4).to_DM(field=True)

In [27]: L, U, _ = M.lu()

In [28]: L*U == M
Out[28]: True

In [29]: M.to_Matrix()
Out[29]: 
⎡75  25  1   47⎤
⎢              ⎥
⎢33  76  54  17⎥
⎢              ⎥
⎣34  56  80  34⎦

In [30]: L.to_Matrix()
Out[30]: 
⎡1    0   0⎤
⎢          ⎥
⎢11        ⎥
⎢──   1   0⎥
⎢25        ⎥
⎢          ⎥
⎢34  134   ⎥
⎢──  ───  1⎥
⎣75  195   ⎦

In [31]: U.to_Matrix()
Out[31]: 
⎡75  25    1     47  ⎤
⎢                    ⎥
⎢        1339   -92  ⎥
⎢0   65  ────   ──── ⎥
⎢         25     25  ⎥
⎢                    ⎥
⎢        16028  24736⎥
⎢0   0   ─────  ─────⎥
⎣         375   1625 ⎦
```

oscarbenjamin commented:
I'm happy to hold off if you're still pushing changes but otherwise I'd say it is good to merge if you're finished.

isuruf commented:
I'm done with the changes.

oscarbenjamin commented:
Okay looks good. Thanks Isuru!
